### PR TITLE
Update .asf.yaml to avoid merge PRs without approvals to the active release branches

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

We create PRs for the release branches once conflicts are not easy to handle when cherry-picking PRs.
Using a PR to resolve the conflicts and get more eyes on the changes to the release branches to reduce the risks that are introduced to the release branches.

But the approvals are not enforced for the release branches when merging PRs so that the committer can merge the PR without any approvals to the release branch.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: .asf.yaml
Comment (tisonkun) [inline .asf.yaml:83]:
```suggestion
```

ditto others.

We don't have CODEOWNER and `dismiss_stale_reviews: false` is the default.

Comment (tisonkun) [reply .asf.yaml:83]: Also I think 2.7 should be end of life once 2.11 released?

**Review by nicoloboschi — COMMENTED** [review]
Is it still possible to directly cherry pick without passing from a pull request?

**Comment by tisonkun** [comment]
@nicoloboschi IIUC `required_pull_request_reviews` infers `Require a pull request before merging` thus it's impossible after this change gets merged.

**Comment by codelipenghui** [comment]
> Is it still possible to directly cherry pick without passing from a pull request?

I think so, the change should only affect the new pull request to release branches.

File: .asf.yaml
Comment (codelipenghui) [reply .asf.yaml:83]: > Also I think 2.7 should be end of life once 2.11 released?

Maybe we will not have a release for branch-2.7. But the PR cherry-picking can also happen on branch-2.7. As I know, there are many users still using 2.7.x

**Comment by nicoloboschi** [comment]
> @nicoloboschi IIUC `required_pull_request_reviews` infers `Require a pull request before merging` thus it's impossible after this change gets merged.

yes that's what I thought. There's no difference in the new config with the master branch so I guess it wouldn't be possible anymore.

If so, we cannot enforce for every cherry-pick to open a pull request, it would create a big noise in the repository without value.
I think we can just advise on the mailing list to wait (and ask!) for a review in case of hard conflicts, for the sake of the project itself

**Comment by codelipenghui** [comment]
@nicoloboschi @tisonkun We will not enforce for every cherry-pick to open a pull request. The committers can push the commit to the release branches after this change. It is to avoid someone opening a PR to release branches but merging it without approvals(The PR can from cherry-pick, or direct fix by committers or contributors)

**Comment by tisonkun** [comment]
> IIUC `required_pull_request_reviews` infers `Require a pull request before merging`

@codelipenghui you can try merge this patch and see if you can direct push to those branches still.

**Comment by codelipenghui** [comment]
image

I have tried it on my fork repo, but I can push the commit to the master branch directly

https://github.com/codelipenghui/incubator-pulsar/commits/master

I'm not sure if I missed something.

**Comment by tisonkun** [comment]
@codelipenghui Such config change is not one-way effect so you can always try and revert if something unexpected :)

**Comment by codelipenghui** [comment]
> Such config change is not one-way effect so you can always try and revert if something unexpected :)

Got it. Yes, if we are not able to cherry-pick directly. We can revert this PR.

File: .asf.yaml
Comment (codelipenghui) [reply .asf.yaml]: @tisonkun addressed. Please check again.

**Review by tisonkun — APPROVED** [review]
Unblock. We can give it a try.

**Review by RobertIndie — APPROVED** [review]
+1. Let's give it a try.

**Comment by codelipenghui** [comment]
Bad news:

```
~/GitHub/apache/incubator-pulsar (branch-2.11) » git push apache branch-2.11                                                                                             lipenghui@lipenghuideMacBook-Pro-2
Enumerating objects: 39, done.
Counting objects: 100% (39/39), done.
Delta compression using up to 12 threads
Compressing objects: 100% (15/15), done.
Writing objects: 100% (22/22), 5.41 KiB | 2.70 MiB/s, done.
Total 22 (delta 5), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/branch-2.11.
remote: error: At least 1 approving review is required by reviewers with write access.
To https://github.com/apache/pulsar.git
 ! [remote rejected]         branch-2.11 -> branch-2.11 (protected branch hook declined)
error: failed to push some refs to 'https://github.com/apache/pulsar.git'
```

After merging this PR, we cannot push the cherry-picked commit directly. I will revert it.
Interesting, why changing the configuration via the GitHub Settings works.

**Comment by tisonkun** [comment]
> Interesting, why changing the configuration via the GitHub Settings works.

@codelipenghui It's possible that you're the admin of your fork and bypass the protect rules.

**Comment by dave2wave** [comment]
If you want to discuss this ASF Infra then I would recommend #asfinfra on the asf slack.

**Comment by tisonkun** [comment]
@codelipenghui @nicoloboschi Today I'm generating release notes for 2.9.4 and noticed that if we directly push a commit to the cherry-picked  branch, it can get missed in the release notes (if generating via GitHub release).

**Comment by tisonkun** [comment]
So requiring PRs for cherry-picking may be a good idea. Also, it can get more visibility and reviews. We do fix a few cherry-picking issues after directly pushing the commit without running regression tests.

**Comment by tisonkun** [comment]
In the Flink community, we require PRs for cherry-picks and it doesn't slow the traffic.

**Comment by tisonkun** [comment]
But we may use `git log --format=oneline v2.9.3...v2.9.4` instead of `gh release view "v2.10.2" -R apache/pulsar --json body --jq .body`. Let me think of it...

**Pull request merged**
